### PR TITLE
Add supply traversals for borders for common allies of both suppliers

### PR DESCRIPTION
**Pull request by agrrr3**
in order to fix transitive use of allied supply this adds supply traversals for borders for common allies of both suppliers

PHASE1: make it work (probably done)
* [x] check logs CHECK BORDERS, supply traversals after ally connections, supply traversals after merging allies

PHASE2: make it right
* [ ] functional code review (does this introduce bugs)
* [~] thorough code review 
* [~] maybe code needs a bit clean up
* [~] maybe better debug output
* [~] change stringtables 

PHASE3: make it fast (probably not in this PR)
* [~] probably can be sped up moving code into the previous loop
* [~] probably can be sped up by removing visited empires

@geoffthemedio can you have a look if I broke anything? i would like to merge a working version before the weekly build; i suggest I fix secondary concerns (style, refactoring for speed, ...) in a followup-PR if the time runs out

Note: as a follow-up there is also a discussion if traversal between warring empires should be allowed or not. but that needs some more thought (e.g. consistency with military ships etc) before introducing extra case switches.

### CODE REVIEW

**Comment by geoffthemedio** [comment]
I don't have time to fully check this, but the thing I'd verify is that, if adding supply connections to empire A's network between systems supplied by empires B and C, where (A and B) and (A and C) are allied but (B and C) are not allied, it should not result in B or C getting that same supply connection added to their network after merging A's connections into B and C's connections.

File: Empire/Supply.cpp [hunk 1]
```diff
@@ -822,6 +822,49 @@ void SupplyManager::Update(const ScriptingContext& context) {
...
     }
+
+
+    for (const auto& [supply_empire_id, empire] : empires) {
+    //for (const auto& [supply_empire_id, empire_obstructed_traversals] : m_supply_starlane_obstructed_traversals) {
+        //for (empire_id : empires.GetEmpireIDsWithDiplomaticStatusWithEmpire(DiplomaticStatus::DIPLO_ALLIED)) {
+        const auto allies_of_empire = allies_of(supply_empire_id);
+        if (allies_of_empire.empty())
+            continue;
+        auto& empire_supply_traversals = ally_merged_supply_starlane_traversals[supply_empire_id];
```
Comment (geoffthemedio) [inline Empire/Supply.cpp]: unused reference?

File: Empire/Supply.cpp [hunk 2]
```diff
@@ -822,6 +822,49 @@ void SupplyManager::Update(const ScriptingContext& context) {
...
+        const auto& supply_empire_baseline_unobstructed_systems =
```
Comment (geoffthemedio) [inline Empire/Supply.cpp]: also unused

**Comment by agrrr3** [comment]
> I don't have time to fully check this, but the thing I'd verify is that, if adding supply connections to empire A's network between systems supplied by empires B and C, where (A and B) and (A and C) are allied but (B and C) are not allied, it should not result in B or C getting that same supply connection added to their network after merging A's connections into B and C's connections.

i verified that (in the test game i uploaded) by having a look at the resulting supply connections etc... in the log. empire 1 is allied to both empires 2 and 3. empire 2 and 3 are at war. the universe is like empire-3, empire-2, empire-1. the result is as expected, empire 1 has a single supply group containing all systems. empire 2 has a single supply group containing empire-1 and empire-2. empire 3 has two supply groups, one containing empire-1 and one containing empire-3

side-note: the crucial thing missing before was the border between 230 (empire-3) and 270 (empire-2).
```
23:00:05.433495 {0x00007fbb8ff1b780} [debug] supply : Supply.cpp:1016 : Connected supply groups for empire 1:
23:00:05.433505 {0x00007fbb8ff1b780} [debug] supply : Supply.cpp:1018 :  ... 0, 25, 45, 55, 70, 90, 105, 130, 135, 140, 160, 205, 220, 230, 235, 240, 245, 265, 270, 275, 
23:00:05.433520 {0x00007fbb8ff1b780} [debug] supply : Supply.cpp:1016 : Connected supply groups for empire 2:
23:00:05.433529 {0x00007fbb8ff1b780} [debug] supply : Supply.cpp:1018 :  ... 0, 25, 90, 105, 130, 135, 140, 160, 205, 220, 235, 240, 245, 265, 270, 275, 
23:00:05.433542 {0x00007fbb8ff1b780} [debug] supply : Supply.cpp:1016 : Connected supply groups for empire 3:
23:00:05.433550 {0x00007fbb8ff1b780} [debug] supply : Supply.cpp:1018 :  ... 25, 105, 130, 135, 140, 205, 235, 240, 265, 
23:00:05.433562 {0x00007fbb8ff1b780} [debug] supply : Supply.cpp:1018 :  ... 45, 55, 70, 230, 
```

edit: fix evil typo: empire 2 has a single supply group containing empire-2 and empire-3.  -> empire-1 and empire-2.

**Comment by geoffthemedio** [comment]
> empire 2 and 3 are at war. [...] empire 2 has a single supply group containing empire-2 and empire-3.

That doesn't sound right...

**Comment by agrrr3** [comment]
> > empire 2 and 3 are at war. [...] empire 2 has a single supply group containing empire-2 and empire-3.
> 
> That doesn't sound right...

yeah, typo at the worst possible place. should have written: single supply group containing empire-1 and empire-2.  (and thats what is the output). With what i wrote one can see that empire 3 and empire 2 do not share any supplied systems besides empire-1. Anyway, here the supplied systems
```
23:00:05.428356 {0x00007fbb8ff1b780} [trace] supply : Supply.cpp:711 : SupplyManager::Update: after removing conflicts, empires can provide supply to the following system ids (and ranges in jumps):
23:00:05.428366 {0x00007fbb8ff1b780} [trace] supply : Supply.cpp:713 :  ... empire 1:  25 (3),  105 (2),  130 (1),  135 (0),  140 (1),  205 (0),  235 (1),  240 (1),  265 (2),  
23:00:05.428395 {0x00007fbb8ff1b780} [trace] supply : Supply.cpp:713 :  ... empire 2:  0 (1),  90 (1),  160 (3),  220 (0),  245 (1),  270 (1),  275 (2),  
23:00:05.428418 {0x00007fbb8ff1b780} [trace] supply : Supply.cpp:713 :  ... empire 3:  45 (0),  55 (3),  70 (2),  230 (1), 
```

**Pull request merged**